### PR TITLE
Support for async actions (redux-thunk)

### PR DESCRIPTION
**Pull request by cvanleeuwen**
I was trying out the bindings in combination with the [redux-thunk middleware](https://github.com/gaearon/redux-thunk) which allows do async actions ([more](http://redux.js.org/docs/advanced/AsyncActions.html)).

What I did was remove the feature to dispatch a polymer function which is called by polymer-redux to produce an action. I deemed it best to leave dispatch handling to the redux dispatcher, and treat the named actions as an interceptor/enhancement on top of redux.

Let me know if this works for you!

Ps. great work on these bindings! Thanks!

### CODE REVIEW

**Comment by tur-nr** [comment]
I'm open to revisit the `dispatch` API on polymer components. I think I just need an example of why it would be beneficial to remove the function call from the `dispatch` method on `PolymerRedux` components.

So take the example of the async reducer from the redux documentation you linked to, will this PR achieve the following:

```javascript
Polymer({
    foo() {
        this.dispatch((dispatch) => {
            dispatch(requestPosts(subreddit));
            return fetch(`http://www.reddit.com/r/${subreddit}.json`)
                .then(response => response.json())
                .then(json => dispatch(receivePosts(subreddit, json)));
        });
    }
});
```

Or if not, can you please demo an async reducer along with a component?

Cheers :smile:

**Comment by cvanleeuwen** [comment]
Your example is spot on!

What was your use-case for the existing behavior (dispatching an function w/o arguments)?

We could also be 'smart' about this; e.g. when the function has zero arguments, call it, otherwise just pass it through to the redux dispatcher.

**Comment by tur-nr** [comment]
Well the reason I put it there was for flexibility. Maybe a component needed to compute an action the same as the `actions` property does, but the developer didn't want to use that pattern. Plus it's also good for inversion of control, passing a component an action to dispatch.

I like the idea of `dispatch` being polymorphic, so checking for the functions length could be good.

**Comment by tur-nr** [comment]
Moving forward, let's add the function call back in but do a length check. If it's zero, immediately call and use returning value as action. If not zero, pass the function to redux dispatch to handle.

Can you make the changes please, thanks.

**Comment by cvanleeuwen** [comment]
Done!

File: polymer-redux.js [hunk 1]
```diff
@@ -71,17 +71,13 @@
...
                 }
 
-                // object
-                if (args[0] != null && typeof args[0] === 'object') {
-                    return store.dispatch(args[0]);
-                }
-
-                throw new TypeError('Polymer Redux: <' + tag + '>.dispatch must be given an action name, function or native redux action object');
+                // action
+                return store.dispatch(args);
```
Comment (tur-nr) [inline polymer-redux.js]: This will dispatch an array, not a function? Should it be:

```javascript
return store.dispatch(args[0]);
```

**Comment by cvanleeuwen** [comment]
Ah yes, I figured it would pass them as separate props :hankey:

:tada: Fixed!

**Comment by tur-nr** [comment]
Great work dude :+1:

**Comment by kristfal** [comment]
@tur-nr This PR was nice, but I have a small issue with it. The async `dispatch` method does not have access to the `actions` properties within an element. This causes unneeded inconsistency between async and sync actions.

Ideally, this should work:

```js
Polymer({
    actions: {
        requestPosts(subreddit): {
            return {
                type: 'REQUEST_POSTS',
                subreddit: subreddit
            };
        }
    }
    getSubreddit: function(subreddit) {
        this.dispatch((dispatch) => {
            dispatch(requestPosts(subreddit));
            return fetch(`http://www.reddit.com/r/${subreddit}.json`)
                .then(response => response.json())
                .then(console.log(response));
        });
    }
});
```

However, right now, the `dispatch` method will return an error as it is unable to read `actions`.

**Comment by tur-nr** [comment]
Yes I see the issue here, but just to clarify shouldn't it be the following:

```javascript
this.dispatch((dispatch) => {
    dispatch(this.requestPosts(subreddit); // notice this.
});
```

Even if that did work, you are using the redux dispatcher and not the polymer-redux dispatch. So maybe it would be best to pass the polymer-redux dispatch to the async function instead. I will take a look tomorrow. Thanks.

**Comment by kristfal** [comment]
@tur-nr Thanks, you're right. I mixed up the example. To follow the sync pattern, shouldn't it be something like this?

```js
Polymer({
    actions: {
        setName: function(first, last) {
            return {
                type: 'SET_NAME',
                first: first,
                last: last
            };
        }
    },
    handleClick: function() {
        this.dispatch((dispatch) => {
          dispatch('setName', 'James', 'Bond');
        }
    }
});
```

**Pull request merged**